### DISCUSSION TITLE
Should Source Generators be used as a replacement for class libraries

### DISCUSSION THREAD

**Discussion by TwentyFourMinutes** [dotnet/roslyn#53961] [Q&A]
Let me try to clarify my question with an example. Assume a library of some sort which has a built in logging system and it is published as `Library.Core`. As the maintainer you want to add a wrapper around your logging system to add support for `Microsoft.Extensions.Logging`. You create a working implementation and publish it as a new package under the name `Library.Extensions.Logging` in order to prevent the need for a dependency to `Microsoft.Extensions.Logging` in your `Library.Core` package. 

Often such implementations are just a couple lines of boilerplate code and at least in my opinion, it feels like unnecessary to create a new package for such little code.

The main problem really is, that you would need to reference `Microsoft.Extensions.Logging`, just for an optional feature. However, back to the example. 

With the introduction of Source Generators you could add the code inside of `Library.Extensions.Logging` to a SG which is being shipped with `Library.Core`. Furthermore, you could generate the wrapper code only if the target assembly references `Microsoft.Extensions.Logging`.

This works fairly well in practice, however is it a _good_ practice? 
I am not quite sure if this is the right place to ask such a question, so feel free to redirect me.

**Comment by MisinformedDNA** [comment]
Regardless of how much code it takes, developers are used to packages that are neatly divided, especially since many are cognizant of their dependencies and this would essentially be sneaking the dependency in. This might take some developers by surprise, may cause version conflicts or legal issues.

**Comment by PathogenDavid** [comment]
If your only goal is to add static C# code to a consuming project, you could already do that without source generators via [MSBuild props/targets embedded in the NuGet package](https://docs.microsoft.com/en-us/nuget/concepts/msbuild-props-and-targets).

Unless this is solving other technical challenges too, this mostly feels like a quest to cut down on the number of assemblies in the output directory. If your consumers care about that you should just let them opt in to something like single-file publishing instead.

**Reply by jasonmalinowski** [reply]
~~You don't even need to make MSBuild props/targets, there's support in NuGet for "content files" which means it'll add the files as Compile items for you: https://docs.microsoft.com/en-us/nuget/reference/nuspec#including-content-files~~

Never mind, I missed the "this is conditioned on the other assembly" bit at first.
